Adding URPF-1.1: uRPF validation from non-default network-instance test

keysightgems opened the pull request:
**Readme Location:** https://github.com/openconfig/featureprofiles/blob/main/feature/urpf/otg_tests/urpf_vrf_test/README.md

**Find the attached logs here:** https://partnerissuetracker.corp.google.com/u/1/issues/415458482

dplore commented:
/fptest virtual

dplore commented on feature/urpf/otg_tests/urpf_vrf_test/urpf_vrf_test.go:
```diff
@@ -0,0 +1,559 @@
+// configureRouteLeaking configure/delete route leaking.
+func configureRouteLeaking(t *testing.T, dut *ondatra.DUTDevice, enable bool) {
+	t.Helper()
+	if deviations.NetworkInstanceImportExportPolicyOCUnsupported(dut) {
```
Move the configuration into a cfgplugin function named after the OC named feature such as `NewInterInstancePolicy`

Then inside `NewInterInstancePolicy` place the deviation which can generate the CLI instead of the OC.

See reference https://github.com/openconfig/featureprofiles/blob/main/internal/cfgplugins/README.md

dplore commented on internal/cfgplugins/staticnexthop.go:
```diff
@@ -217,3 +229,99 @@ func NextHopGroupConfigForMulticloud(t *testing.T, dut *ondatra.DUTDevice, traff
+}
+
+// configureUDPEncapHeader configure UDP encapsulation header.
+func configureUDPEncapHeader(t *testing.T, dut *ondatra.DUTDevice, tunnelType string, dstPort uint16) {
```
this function appears unused? If yes, please remove

dplore commented on internal/cfgplugins/policyforwarding.go:
```diff
@@ -878,3 +889,91 @@ func ConfigureDutWithGueDecap(t *testing.T, dut *ondatra.DUTDevice, guePort int,
+}
+
+// InterfacePolicyForwardingApply configure to apply policy forwarding.
+func InterfacePolicyForwardingApply(t *testing.T, dut *ondatra.DUTDevice, interfaceName string, policyName string, ni *oc.NetworkInstance, params OcPolicyForwardingParams) {
```
If keeping this function, then Exported cfgplugins functions must use parameters formatted per:
https://github.com/openconfig/featureprofiles/blob/main/internal/cfgplugins/README.md#function-parameters

keysightgems commented:
@dplore 
I have addressed all the above comments and committed the code.
Attached the latest logs here https://partnerissuetracker.corp.google.com/u/1/issues/415458482

dplore commented on internal/cfgplugins/staticnexthop.go:
```diff
@@ -275,28 +278,29 @@ func NextHopGroupConfigForIpOverUdp(t *testing.T, dut *ondatra.DUTDevice, ni *oc
 			}
 
-			if deleteTtl {
+			if params.DeleteTtl {
```
Please add a comment explaining what this does.

dplore commented on internal/cfgplugins/staticnexthop.go:
```diff
@@ -172,15 +172,18 @@ type StaticNextHopGroupParams struct {
+	TTL                uint8
+	DSCP               uint8
+	NetworkInstanceObj *oc.NetworkInstance
+	DeleteTtl          bool
```
Please add a comment explaining what the purpose of this var is

dplore commented on internal/cfgplugins/policyforwarding.go:
```diff
@@ -51,15 +51,18 @@ type OcPolicyForwardingParams struct {
+	TunnelIP           string
+	InterfaceName      string
+	PolicyName         string
+	NetworkInstanceObj *oc.NetworkInstance
```
add a comment explaining what the purpose of this var is / how it is used

dplore commented on internal/cfgplugins/policyforwarding.go:
```diff
@@ -73,6 +76,17 @@ type PolicyForwardingRule struct {
 }
 
+type GueEncapPolicyParams struct {
+	TrafficType      string
```
add comment explaining the purpose of this var

dplore commented on internal/cfgplugins/policyforwarding.go:
```diff
@@ -73,6 +76,17 @@ type PolicyForwardingRule struct {
+	DstAddr          []string
+	SrcAddr          []string
+	Ttl              uint8
+	Rule             uint8
```
add comment explaining the purpose of this var

dplore commented on internal/cfgplugins/staticnexthop.go:
```diff
@@ -171,6 +172,20 @@ type StaticNextHopGroupParams struct {
 
+// NexthopGroupUDPParams defines the parameters used to create or configure a Next Hop Group that performs UDP encapsulation for traffic forwarding.
+type NexthopGroupUDPParams struct {
+	TrafficType        string
```
Add comment explaining purpose of this var.  Looks like it is used for IP address family?  Maybe rename it so it is more specific

dplore commented on internal/cfgplugins/staticnexthop.go:
```diff
@@ -217,3 +232,84 @@ func NextHopGroupConfigForMulticloud(t *testing.T, dut *ondatra.DUTDevice, traff
+			case "V6Udp":
+				groupType = "ipv6-over-udp"
+			default:
+				t.Fatalf("Unsupported traffic type %q", params.TrafficType)
```
```suggestion
				t.Fatalf("Unsupported address family type %q", params.TrafficType)
```

keysightgems commented:
@dplore 
Addressed all the comments and fixed. Please check and approve. Thanks!
Logs attached here https://partnerissuetracker.corp.google.com/u/1/issues/415458482

keysightgems commented on feature/urpf/otg_tests/urpf_vrf_test/urpf_vrf_test.go:
```diff
@@ -0,0 +1,550 @@
+				}
+			}
+			if !tc.expectLoss {
+				configureRouteLeaking(t, dut, batch, true)
```
Removed the route leaking configuration and added static routes as per the README.
Find the latest test validation logs here https://partnerissuetracker.corp.google.com/u/1/issues/415458482.

dplore replied:
Thank you!

keysightgems commented on internal/cfgplugins/bgp_policy.go:
```diff
@@ -94,3 +96,46 @@ match as-path %s
 }
+
+// ConfigureRouteLeakingFromCLI configure route leaking through CLI.
+func ConfigureRouteLeakingFromCLI(t *testing.T, dut *ondatra.DUTDevice, vrfName string) {
```
Yes, I tried using the OC path, but encountered an error. I added a deviation and handled this case directly in the code. (Deviation: https://partnerissuetracker.corp.google.com/u/1/issues/457884385)

dplore reviewed (approved):
LGTM, but there are some conflicts to resolve
